### ISSUE TITLE
how difficult will it be for developers to port projects / general questions

### ISSUE THREAD

**Issue by ohitstom** [Delusoire/bespoke#4]
as a developer how different will the bespoke and spicetify api's be? for example will all of the react apis stay the same in naming and what kind of direction is being taken with cosmos and platform methods? will css mappings stay the same?

im also worried about performance, currently it takes a good 2-3 seconds to load bespoke on spotify for me where-as spicetify is instant as its patched files i assume - it feels like bespoke itself has the same amount of delay as marketplace does when loading themes within spicetify and it also breaks spotifys default colors / layout during this 2-3 second span which looks super weird.

i understand this is still work in progress but will a full release automatically enable all of your own extensions because that would be really annoying for the end user but im assuming this is just because of development reasons atm.

just wondering what the end goal in this taking over spicetify will be - currently it seems like a downgrade(?) to the end user apart from the new slick marketplace design and code quality for devs.

also will developers have to put their projects in your bespoke-modules repo because that seems less than ideal.

also will spicetify still keep its name or will bespoke become its new brand because that seems silly seeing as there are already popular repos with the same name!

just worried on the future of my projects as a developer not trying to be negative of bespoke at all haha!

**Comment by Delusoire** [comment]
1. With every major release, breaking changes are introduced. You can't expect extensions to work as is when migrating from v2 to v3. The general direction is to stick as much as possible to the Spotify methods and naming conventions. Platform is exposed as is, and Cosmos is extracted from Platform and exposed too. If by Cosmos you mean the internal protocols then the trend is for Spotify to deprecate them and replace them with Platform APIs.

2. I understand that performance may be sub-par for lower-end hardware, bespoke patches Spotify at runtime and I'm not currently doing anything to optimize this process. I can think of so many ways to improve the performance of this like building a dependency graph and loading modules concurrently instead of sequentially, because some extensions like library-db and search-on-youtube need to do a lot of work upon initialization.

3. Yeah

4. The cli is easier to use, backups never go into an erroneous state, mono manifest for marketplace is finally implemented, modules are installed locally, and you can choose to run bespoke without tampering with the Spotify installation (the --mirror switch). Also, the dev experience is important; The easier it is to create new modules the better the ecosystem will be and the more high-quality modules will be produced.

5. the new Marketplace offers unmatched flexibility, you can install modules from any repo just by specifying a raw GitHub URL to a metadata.json file. You can have monorepos, meta-modules (modules inside of modules), or any structure that you please. Using GitHub raw links also allows you to pin the version of the module to a specific commit/tag/branch and have it auto-update by marketplace. But for security reasons, the only browsable extensions are those listed inside marketplace's repo.json (mono manifest).

6. Spicetify will keep it's name. bespoke is more like a code name

**Comment by ohitstom** [comment]
awesome so just to clarify wherever possible naming conventions will stay the same (idk something like Spicetify.Config etc) and that I will not need to include my projects in anyone else's repos?

will it be possible for an option to not patch at runtime in future for lower end hardware assuming performance cannot be improved substantially - I think that would be a nice to have if possible

**Comment by Delusoire** [comment]
I don't have a list of changes, but I'll start writing documentation once everything falls into place. `Spicetify.Config` doesn't exist anymore because of the architectural changes.

I'll probably develop a caching system at some point for the mixins but that's not a priority. Also, keep in mind that the only reason modules are taking some time to load is because bespoke, before executing a module, waits for all the module's dependencies to load. Meaning that most modules won't actually execute till the "stdlib" is completely loaded and has exposed all Platform, Cosmos, webpack ... methods. This again can be reduced by splitting "stdlib" into smaller libraries which I'll eventually do too.

By the way, you should be able to notice a speedup in loading time as concurrent module loading has just been implemented.

**Comment by ohitstom** [comment]
> This again can be reduced by splitting "stdlib" into smaller libraries which I'll eventually do too

is there any reason this can't be done how it was before where every spicetify definition was relatively independent already and just kinda loaded whenever it could (within stdlib itself so it doesn't have to be split up)?

also wild thing to ask but any reason why (apart from platform methods) some slower webpack functions could be stored in local storage and updated independently after stdlib has finished loading? kind of like how marketplace works in V2 but maybe that's a stupid thing to ask lol

potentially we could implement a proper dependency system for extension Devs so they don't have to do a bunch of timeout stuff though I'm not sure how that would look

**Comment by Delusoire** [comment]
> is there any reason this can't be done how it was before where every spicetify definition was relatively independent already and just kinda loaded whenever it could (within stdlib itself so it doesn't have to be split up)?

Most things are still exposed independently but they're all exported from the same module. spicetify v2 wasn't loading everything independently, most functions had setTimeouts and setIntervals creating a non linear mess of control flow. bespoke modules are the smallest unit, a module can't depend on a third of another module... I can't see why you could prefer the older approach.

> also wild thing to ask but any reason why (apart from platform methods) some slower webpack functions could be stored in local storage and updated independently after stdlib has finished loading? kind of like how marketplace works in V2 but maybe that's a stupid thing to ask lol

Can you rephrase your question?

> potentially we could implement a proper dependency system for extension Devs so they don't have to do a bunch of timeout stuff though I'm not sure how that would look

It is already implemented.

**Comment by ohitstom** [comment]
> Can you rephrase your question?
I guess what I mean is once a function is exposed through webpack, idk like a certain react component, we cache it in local storage and on loadup use that cached function initially so that we have instant availability, then once the usual code for getting these webpack definitions is finished it updates the local storage and globalThis definitions?


> It is already implemented.

awesome!! looking forward to using it

oh also is there any way I can use the currently implemented APIs from console without having to import in my own module and then expose to globalThis? or is this planned for future yet

**Comment by ohitstom** [comment]
> Can you rephrase your question?
I guess what I mean is once a function is exposed through webpack, idk like a certain react component, we cache it in local storage and on loadup use that cached function initially so that we have instant availability, then once the usual code for getting these webpack definitions is finished it updates the local storage and globalThis definitions?


> It is already implemented.

awesome!! looking forward to using it

oh also is there any way I can use the currently implemented APIs from console without having to import in my own module and then expose to globalThis? or is this planned for future yet

**Comment by ohitstom** [comment]
> Can you rephrase your question?
I guess what I mean is once a function is exposed through webpack, idk like a certain react component, we cache it in local storage and on loadup use that cached function initially so that we have instant availability, then once the usual code for getting these webpack definitions is finished it updates the local storage and globalThis definitions?


> It is already implemented.

awesome!! looking forward to using it

oh also is there any way I can use the currently implemented APIs from console without having to import in my own module and then expose to globalThis? or is this planned for future yet

**Comment by Delusoire** [comment]
1. That's not possible

2. `const { S } = await import("/modules/Delusoire/stdlib/index.js")`

3. Probably not, unless you want to make a PR for it 😉

4. Rely on what? I can't see any use for it.

**Comment by ohitstom** [comment]
> 1. That's not possible

fair enough 

> 2. `const { S } = await import("/modules/Delusoire/stdlib/index.js")`

perfect, I'm guessing we aren't removing the global Spicetify definition on full release though right?!

> 3. Probably not, unless you want to make a PR for it 😉

haha could do, not sure how this could be done on Mac but Linux and windows seems simple enough 


> 4. Rely on what? I can't see any use for it.

many extensions use it to reference spicetify and Spotify versioning for implementing certain features, the Nord and comfy theme both use these in their extensions and are some of the most popular themes!

it's very helpful for adding fallback support (I understand user agent is a thing but spicetify.config is also used in the about section of spotifys version menu which makes troubleshooting with users way easier)

surely this won't be very difficult isn't the new config thing a json?

**Comment by Delusoire** [comment]
1. Of course we are

2. That would be good enough for me, Mac support is already not so great (custom URI schemes on macOS require packaging the executable as a .app ...)

3. You can use `S.Platform.version`

**Comment by ohitstom** [comment]
> 3. You can use `S.Platform.version`

what about current theme, spicetify version, etc? we can't really remove a feature that's used - even marketplace uses it in spots and the wrapper itself (I know you've probably removed it since rewrite but it could come in handy in future)

I don't want to argue but without this feature it would fully break the comfy theme in certain respects - since we rely on knowing the current local colorscheme / colorscheme enabled in marketplace (since it overwrites it's value)

**Comment by Delusoire** [comment]
There is no concept of "current theme", people should be able to mix and match features from different themes as they wish. Also, there is no "spicetify version", there is only a concept of versions for modules, you can't directly access it, but if for some god forbidden reason you find yourself in an obligation to use it then you can fetch the module's metadata.json. Also, if you're trying to prove a point then it would be better to provide an example or two that illustrates it.

**Comment by ohitstom** [comment]
why are you being hostile? I've not been hostile to you, I can't provide examples as I'm not at a computer right now. I wasn't trying to prove a point I was just giving examples of where it's currently used and why it would be a shame if we lost that functionality, it will make it a lot harder for developers to support V2 and v3 at the same time if we cannot know what version of Spicetify is being used.

I don't really understand how themes will be mixed and matched that sounds like a recipe for disaster, I will just wait and see what happens on that front I guess. if we are removing command line configuration for themes or whatever that will suck completely for a lot of users especially Linux users but I assume that's not what you mean.

some sort of global variable exposing the current spicetify version would still be handy at the very least please `Spicetify.Version` or something

**Comment by Delusoire** [comment]
I'm not trying to sound hostile.
All recent spicetify releases have been bug fixes and patches.
Most themes in the old marketplace are somewhat bloated, you can easily split them into a theme, extensions and snippets. I understand that you don't have access to the Discord server and as such haven't really been part of the conversation, so here's the primer:
bespoke itself is very minimal and doesn't expose any of the Spotify internals. It's basically a module loader, modules are the smallest unit of loadable object, they can be enabled, disabled, installed and removed at runtime without reloading the client. Modules have 3 entry points as specified by the metadata.json; A CSS entry, A JS entry and a Mixins entry. Mixins are basically javascript but they are executed before Spotify is loaded and can directly patch Spotify with regex search and replace. Modules can also depend on each other, thereby increasing code reusability. Bespoke also forces you to use es modules (no global variables) and encourages you to organize your code into different files. Bonus point: tailwind support is enabled by default!
Now stdlib is just a normal module like any other, but as of now, it is functionally equivalent to the old CLI's preprocess.go and the spicetifyWrapper.js. All themes, snippets, extensions and custom apps are just modules too (this implies that custom apps can be installed enabled, and used at runtime too). All these operations can be done through the marketplace interface but they ultimately are processed at the CLI level through a custom URI scheme. Meaning all customization can be done through the CLI.

Now back to Spicetify.Config, why is the Spotify version not enough? Why would you need the stdlib version? Maybe i'll eventually add the ability to get the versions of all the modules that the current module is depending on.

**Comment by ohitstom** [comment]
ah that makes a lot more sense, I'm not sure what forcing Devs to use es stuff entails but hopefully nothing too crazy, I assume we won't be forced to use Spicetify creator or similar

as for spicetify version it's because specific extensions use this when supporting alot of Spotify versions, for example where <2.13 used a cosmos request but >2.20 uses a platform method or graphql call for instance.

it also means for some of my projects I can differentiate if the user is on v1 v2 or V3 and load the correct CSS / JS for that version of spicetify, it also means I can disable certain settings if they are unsupported on the users current Spotify and spicetify version.

I understand this may be niche but I can't see any other way of handling this if this feature is removed other than a bunch of try except or other error checking stuff.

maybe I'm misunderstanding of course but that is my main concern, in comfy we also use the config for knowimg the current colorscheme so we don't overwrite it when the user is using our remote fetched schemes (changing schemes in our settings menu which may become obsolete in v3 if theres no explicit way of enabling schemes at the cli level and instead in marketplace)

also what does it mean by splitting up themes into extensions and snippets etc are we not sticking to the former setup of:
CSS file
js file
assets folder

I know alot of these questions will be answered in docs eventually I just want to be slightly ahead of the curve so I can plan future support as the only Dev of one of the big-ish actively maintained themes still on marketplace... (excluding yours of course)

**Comment by Delusoire** [comment]
1. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules
You can use plain js

2. Wouldn't it make more sense to use the Spotify version for that?

3. Ain't no one using v1 of spicetify. Also once v3 will be released there won't be any support for v2.

4. Ultimately, your module has the power to do whatever it wants and change Spotify internals in any way it wants, you're not limited by stdlib.

5. I'm saying that theme developers should refrain from adding everything to a single package, a lot of the themes are doing similar things but in implemented differently. That's just horrible.

**Comment by ohitstom** [comment]
> 1. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules
> You can use plain js
> 
> 2. Wouldn't it make more sense to use the Spotify version for that?
> 
> 3. Ain't no one using v1 of spicetify. Also once v3 will be released there won't be any support for v2.
> 
> 4. Ultimately, your module has the power to do whatever it wants and change Spotify internals in any way it wants, you're not limited by stdlib.
> 
> 5. I'm saying that theme developers should refrain from adding everything to a single package, a lot of the themes are doing similar things but in implemented differently. That's just horrible.

awesome thanks for all the info, any thoughts on enabling discussions for this repo ?

**Comment by Delusoire** [comment]
I'll enable discussions
The new marketplace supports selecting versions based on branch, tag and commit.
I'm on my last percent of battery, rip.

**Comment by ohitstom** [comment]
> I'll enable discussions
gotcha thank you, I would still like to potentially discuss adding the spice version and ignoring the rest of spicetify.config but that can be in another issues or discussions
